File: loc2ip/loc2ip.py
#!/usr/bin/python

import sys

def main():

    try:
        #sys.stdout.write("Enter Latitude in degrees [-90, 90]: ")
        lat = sys.stdin.readline()
        lat = float(lat)
        if lat > 90 or lat < -90:
            sys.stderr.write(sys.argv[0]+": Latitude out of bounds. [-90, 90]\n")
            print("0000:0000:0000:0000:0000:0000:0000:0000")
            exit(1)

        #sys.stdout.write("Enter Longitude in degrees [-180, 180]: ")
        long = sys.stdin.readline()
        long = float(long)
        if long > 180 or long < -180:
            sys.stderr.write(sys.argv[0]+": Longitude out of bounds. [-180, 180]\n")
            print("0000:0000:0000:0000:0000:0000:0000:0000")
            exit(1)

        #sys.stdout.write("Enter Altitude in metres [-100, 6453.6): ")
        alt = sys.stdin.readline()
        alt = float(alt)
        if alt < -100 or alt >= 6453.6:
            sys.stderr.write(sys.argv[0]+": Altitudes of this magnitude are unsupported. [-100, 6453.6)\n")
            print("0000:0000:0000:0000:0000:0000:0000:0000")
            exit(1)
    except ValueError:
        sys.stderr.write(sys.argv[0]+": Arguments must be of type float.\n")
        print("0000:0000:0000:0000:0000:0000:0000:0000")
        exit(1)

    lat += 90
    lat /= 180
    lat *= (2**32)-1
    long += 180
    long /= 360
    long *= (2**32)-1
    alt += 100
    alt *= 10
    lat = int(lat)
    long = int(long)
    alt = int(alt)
    print "3ffe:%04x:%04x:%04x:%04x:%04x:dead:beef" % \
            ((lat&(0xffff<<16))>>16, lat&0xffff, \
            (long&(0xffff<<16))>>16, long&0xffff, \
            alt&0xffff)

main()

File: loc2ip/loc2ip.c
#include <stdio.h>
#include <stdlib.h>
#include <tgmath.h>

#define SCRIPTNAME "loc2ip"

void usage() {
    fprintf(stderr, "%s: Generates an IPv6 address, based on geolocation coordinates.\n", SCRIPTNAME);
    fprintf(stderr, "Usage: %s latitude longitude altitude\n", SCRIPTNAME);
    fprintf(stderr, "Example: %s %f %f %d\n", SCRIPTNAME, 40.439167, -79.976667, 40);
}

void invalid_argument(char *arg) {
    printf("Argument invalid\n");

}

int main(int argc, char **argv) {

if (argc != 4) {
    usage();
    exit(1);
}

long longitude = atof(argv[1]);
long latitude = atof(argv[2]);
long altitude = atof(argv[3]);

/* Normalise longitude (0-360) */
if(longitude < 0) {
    longitude += 360;
}

/* Normalise latitude (0-180) */
latitude += 90; 

/* Convert to microseconds */
longitude *=  (360 * 100 * 100);
latitude *= (360 * 100 * 100);


latitude += 90;
latitude /= 180;
latitude *= (2&0xffffffff);
longitude += 180;
longitude /= 360;
longitude *= (2&0xffffffff);
altitude += 100;
altitude *= 10;

/* Convert to integer values */
long longitudeInt = (long) longitude;
long latitudeInt = (long) latitude;
long altitudeInt = (long) latitude;

/*print "3ffe:%04x:%04x:%04x:%04x:%04x:dead:beef" % \
        ((lat&(0xffff<<16))>>16, lat&0xffff, \
        (long&(0xffff<<16))>>16, long&0xffff, \
        alt&0xffff)
*/

/* Output IP Address */
printf("3ffe::%04x::%04x::%04x::%04x::%04x\n", longitudeInt >> 16, longitudeInt & 0xffff,
 latitudeInt >> 16, latitudeInt & 0xffff, (long) altitude);

return 0;

}
